File: src/leetcodejava/string/AddString415.java
package leetcodejava.string;


import org.junit.Assert;
import org.junit.Test;

/**
 * This is the solution of No. 415 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/add-strings
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定两个字符串形式的非负整数 num1 和num2 ，计算它们的和。
 * <p>
 * 注意：
 * <p>
 * num1 和num2 的长度都小于 5100.
 * num1 和num2 都只包含数字 0-9.
 * num1 和num2 都不包含任何前导零。
 * 你不能使用任何內建 BigInteger 库， 也不能直接将输入的字符串转换为整数形式。
 * <p>
 * 来源：力扣（LeetCode）
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class AddString415 {

    @Test
    public void addStringTest() {
        String num1 = "12345";
        String num2 = "12345";
        String num = addStr(num1, num2);
        Assert.assertEquals(num, "24690");
    }

    /**
     * 加字符串
     *
     * @param num1 字符串1
     * @param num2 字符串2
     * @return 字符串和
     */
    public String addStr(String num1, String num2) {
        if (num1 == null || num1.length() < 1) {
            return num2;
        }
        if (num2 == null || num2.length() < 1) {
            return num1;
        }
        StringBuffer sb = new StringBuffer();
        int i = num1.length() - 1;
        int j = num2.length() - 1;
        int carry = 0;
        while (i >= 0 || j >= 0) {
            int total = carry;
            if (i >= 0) {
                total += num1.charAt(i--) - '0';
            }
            if (j >= 0) {
                total += num2.charAt(j--) - '0';
            }
            carry = total / 10;
            sb.append(total % 10);
        }
        if (carry > 0) {
            sb.append(carry);
        }
        return sb.reverse().toString();
    }
}

File: src/leetcodepython/string/add_string_415.py
# encoding='utf-8'

'''
/**
 * This is the solution of No. 415 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/add-binary/
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定两个二进制字符串，返回他们的和（用二进制表示）。
 * <p>
 * 输入为非空字符串且只包含数字 1 和 0。
 * <p>
 * 示例 1:
 * 输入: a = "11", b = "1"
 * 输出: "100"
 * 示例 2:
 * 输入: a = "1010", b = "1011"
 * 输出: "10101"
 * <p>
 * 来源：力扣（LeetCode）
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''


class Solution:
    def add_string(self, num1: str, num2: str) -> str:
        '''
            两个字符串相加
        Args:
            num1:字符串1
            num2:字符串2
        Returns:
            字符串结果
        '''
        sb = []
        i, j, carry = len(num1) - 1, (num2) - 1, 0
        while i >= 0 or j >= 0:
            a = 0 if i < 0 else int(num1[i])
            b = 0 if j < 0 else int(num2[j])
            total = a + b + carry
            sb.append(str(total % 10))
            carry = total // 10
            i -= 1
            j -= 1
        if carry > 0:
            sb.append(str(carry))
        return ''.join(sb[::-1])


if __name__ == '__main__':
    a = "9"
    b = "99"
    solution = Solution()
    result = solution.add_string(a, b)
    print(result)
    assert result == "108"
